### ISSUE TITLE
[Design] Network data aggregator

### ISSUE THREAD

**Issue by r4mmer** [HathorNetwork/hathor-explorer-service#66]
- Feature Name: data_collector
- Start Date: 2021-07-05
- RFC PR: (leave this empty)
- Hathor Issue: (leave this empty)
- Author: Giovane Costa <gigio.coder@gmail.com>

# Summary
[summary]: #summary

Data aggregator will aggregate all network data from Redis, check if something changed and update if necessary.

# Motivation
[motivation]: #motivation

We want to have an overview of all the hathor network. Today we have only a single node point of view. 

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Data aggregator will get data collected by `data-collector`, aggregate them into a Network dataclass, compare with saved data and save it if different.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

Three new dataclasses will be created: `AggregatedNode`, `AggregatedPeer` and `Network`.

- `AggregatedNode` and `AggregatedPeer` will be a bit smaller version of `Node` and `Peer` respectively, having `up_since` as the main change.
  As `uptime` is always changing as time goes by, `up_since` remains the same. 
  Also, `up_since` is an integer to avoid problems with microseconds calculations (`/v1a/status` returns only `uptime` and we have to calculate everytime).
- `AggregatedPeer` also have a list of nodes that it is connected to.
- `Network` have the list of `AggregatedNode` and `AggregatedPeer`

`AggregateNodeData` usecase is called along with `SaveNodeData` inside `aggregate_node_handler`.

`Network` is saved into Redis under the following key:

> `'hathor-explorer-service.network.v1'`

**Endpoint:**

GET `/network`

`network-handler` return current network stored in Redis

```
{
  nodes: [AggregatedNode, ...],
  peers: [AggregatedPeer, ...]
}
```

# Drawbacks
[drawbacks]: #drawbacks

Polling is still bad specially with data that will not be modified constantly. However, WebSocket is intended to be implemented soon.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

We could do the aggregation on client side. However, more requests must be made every second, or always return everyhthing.

The disadvantage of this approach is that we must keep polling. 

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Nodes are still manually setted by us. It would be good to discover new http responding nodes automatically as they join the network.

# Future possibilities
[future-possibilities]: #future-possibilities

WebSocket. Everytime `Network` have new content and is saved, a handler can be triggered to broadcast new data to connected clients.

**Comment by andreabadesso** [comment]
The design is approved by me, I would just add an explanation on the difference between `node` and `peer`

**Comment by r4mmer** [comment]
> ` "state": ("INITIALIZING"|"READY"),`
> 
> Those are really the only possibles states?

Yes, this is from the manager [`NodeState` enum](https://github.com/HathorNetwork/hathor-core/blob/43d0d26e81f489f53b35783d6926972cb329afe4/hathor/manager.py#L60)

**Comment by msbrogli** [comment]
Hi @r4mmer . Here are my points and questions:

1) As far as I understood, Aggregated Nodes are the ones we collect information from, while Aggregated Peers are a list of peers discovered by all nodes. Is that right? If yes, will our nodes appear in the aggregated peers as well? If yes, shouldn't we just have a list of peers with a flag indicating whether it's one of ours or not?

2) How will we present the aggregated information in the public explorer?

3) I feel there are some other possible states, such as `SYNCING`.

**Comment by r4mmer** [comment]
> 1. As far as I understood, Aggregated Nodes are the ones we collect information from, while Aggregated Peers are a list of peers discovered by all nodes. Is that right? If yes, will our nodes appear in the aggregated peers as well? If yes, shouldn't we just have a list of peers with a flag indicating whether it's one of ours or not?

The fields we can gather from a node and a peer differ slightly, so having them on separate lists is better.
But if later we check that these different fields have no utility we can merge the lists.

> 2. How will we present the aggregated information in the public explorer?

The presentation has not been discussed yet, only a few possibilities (like world map and aggregated network info).

> 3. I feel there are some other possible states, such as `SYNCING`.

The `Synchronized` and `Synchronizing` states are computed on the frontend based on the `latest_timestamp` of the server, not on the `state` field from the api.
The `state` from the api is from the [NodeState Enum](https://github.com/HathorNetwork/hathor-core/blob/43d0d26e81f489f53b35783d6926972cb329afe4/hathor/manager.py#L60)
